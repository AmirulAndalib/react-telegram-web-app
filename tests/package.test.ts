import * as fs from 'fs';
import * as path from 'path';

const BUILD_PATH = path.resolve(__dirname, '../lib');

const COMMON_JS_MODULE = 'react-telegram-web-app.cjs';

const REQUIRED_MODULES = [
  COMMON_JS_MODULE,
  'react-telegram-web-app.module.js',
  'react-telegram-web-app.modern.js',
  'react-telegram-web-app.umd.js',
] as const;

const REQUIRED_EXPORTS = [
  'WebAppProvider',
  'MainButton',
  'BackButton',
  'useShowPopup',
  'useHapticFeedback',
  'useThemeParams',
  'useScanQrPopup',
  'useReadTextFromClipboard',
  'useSwitchInlineQuery',
  'useExpand',
];

const REQUIRED_DTS = [
  ...REQUIRED_EXPORTS.map(name => `${name}.d.ts`),
  'useWebApp.d.ts',
  'context.d.ts',
  'index.d.ts',
];

describe('Public API modules', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('checks exist lib modules', () => {
    expect(
      [
        ...REQUIRED_MODULES,
        ...REQUIRED_MODULES.map(module => module + '.map'),
      ].filter(module => !fs.existsSync(path.join(BUILD_PATH, module))),
    ).toStrictEqual([]);
  });

  it('checks exist declaration types', () => {
    expect(
      REQUIRED_DTS.filter(
        module => !fs.existsSync(path.join(BUILD_PATH, module)),
      ),
    ).toStrictEqual([]);
  });

  it('checks export from modules', () => {
    const indexModule = require(path.join(BUILD_PATH, COMMON_JS_MODULE));

    expect(indexModule).toMatchObject(
      REQUIRED_EXPORTS.reduce((memo, name) => {
        memo[name] = expect.any(Function);
        return memo;
      }, {} as Record<string, any>),
    );
  });
});

describe('Public API documentation', () => {
  it('checks contain describe in README.md', () => {
    const mdFile = fs.readFileSync(
      path.resolve(__dirname, '../README.md'),
      'utf8',
    );

    expect(
      REQUIRED_EXPORTS.map(
        name => `[${name}](./docs/README.md#${name.toLowerCase()})`,
      ).filter(name => !mdFile.includes(name)),
    ).toStrictEqual([]);
  });
});
