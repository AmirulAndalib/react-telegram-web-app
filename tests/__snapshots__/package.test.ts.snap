// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package tests /lib snapshot 1`] = `
[
  "/BackButton.d.ts",
  "/MainButton.d.ts",
  "/WebAppProvider.d.ts",
  [
    "/core/context.d.ts",
    "/core/useWebApp.d.ts",
  ],
  "/index.d.ts",
  "/react-telegram-web-app.cjs",
  "/react-telegram-web-app.cjs.map",
  "/react-telegram-web-app.modern.js",
  "/react-telegram-web-app.modern.js.map",
  "/react-telegram-web-app.module.js",
  "/react-telegram-web-app.module.js.map",
  "/react-telegram-web-app.umd.js",
  "/react-telegram-web-app.umd.js.map",
  "/useExpand.d.ts",
  "/useHapticFeedback.d.ts",
  "/useReadTextFromClipboard.d.ts",
  "/useScanQrPopup.d.ts",
  "/useShowPopup.d.ts",
  "/useSwitchInlineQuery.d.ts",
  "/useThemeParams.d.ts",
]
`;
